### PR TITLE
Util/actions

### PR DESCRIPTION
**Pull request by karamba228**

### CODE REVIEW

File: .github/actions/annotate/index.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                if (res.statusCode === 200) {
+                    resolve(JSON.parse(data));
+                } else {
+                    reject(`Request failed with status code ${res.statusCode}`);
+                }
+            });
+        }).on('error', (e) => {
+            reject(`Error: ${e.message}`);
+        });
+    });
```
Comment (srossross) [inline .github/actions/annotate/index.js]: can you replace this with the fetch builtin?
```suggestion
   const res = await fetch(url);
   return res.json()
```

File: .github/actions/annotate/index.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+const fs = require('fs');
+const https = require('https');
+
+// Function to strip versioning from package names
+function stripVersion(packageLine) {
+    return packageLine.split(/[<>=~!]/)[0].trim();
+}
+
+// Function to make an API request using the built-in https module
+function makeApiRequest(packageName) {
```
Comment (srossross) [inline .github/actions/annotate/index.js]: Can you name this to reflect what it is doing `fetchPackageScore`

File: .github/actions/annotate/index.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+async function run() {
+    const filePath = 'requirements.txt';
+
+    if (!fs.existsSync(filePath)) {
+        console.log('::error::requirements.txt not found!');
+        return;
+    }
+
+    const packages = fs.readFileSync(filePath, 'utf-8').split('\n').filter(pkg => pkg);
```
Comment (srossross) [inline .github/actions/annotate/index.js]: Please use the async fs in an async function `const fs = require('fs/promises');`

then 

```suggestion
    const packages = (await fs.readFile(filePath, 'utf-8')).split('\n').filter(pkg => pkg);
```

**Review by srossross — CHANGES_REQUESTED** [review]
A few issues raised

Can you also please use the `package-ecosystem: pip` this should be configurable and raise an error if it is not supported

Also the requirements.txt does not check that the package names are valid eg

* `# This is a comment`
* `package# This is a comment after a package`
* `-i foo.com package`

please see https://pip.pypa.io/en/stable/reference/requirements-file-format/

File: .github/actions/annotate/action.yml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+name: "Annotate Requirements"
+description: "Annotates the requirements.txt file using the Score API"
+runs:
+  using: "node12"
```
Comment (srossross) [inline .github/actions/annotate/action.yml]: why 12? isn't 20 the current LTS version?

File: .github/actions/annotate/index.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            console.log(`::error file=requirements.txt::Package ${packageName} not found.`);
+        }
+    } catch (error) {
+        console.log(`::error file=requirements.txt::Error looking up package ${packageName}: ${error.message}`);
+    }
+}
+
+async function run() {
+    const filePath = 'requirements.txt';
+    const ecosystem = process.env.PACKAGE_ECOSYSTEM || 'pip';
```
Comment (srossross) [inline .github/actions/annotate/index.js]: the example docs I sent say to use 

```ts
const core = require('@actions/core');
core.getInput('package-ecosystem', { required: true });
```

is there a reason you choose not to use that package?

File: .github/actions/annotate/index.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            const { maturity, health_risk } = response.source;
+            const maturityValue = maturity ? maturity.value : 'Unknown';
+            const healthRiskValue = health_risk ? health_risk.value : 'Unknown';
+            
+            console.log(`::notice file=requirements.txt::Package ${packageName} found. (Maturity: ${maturityValue}, Health: ${healthRiskValue})`);
+        } else {
+            console.log(`::error file=requirements.txt::Package ${packageName} not found.`);
+        }
+    } catch (error) {
+        console.log(`::error file=requirements.txt::Error looking up package ${packageName}: ${error.message}`);
```
Comment (srossross) [inline .github/actions/annotate/index.js]: we are trying to answer the question

> Will integrating project X as a dependency enhance the stability and maintainability of our applications while minimizing potential risks and challenges?

in the next PR lets adjust the errors to match the answer to this question better

**Review by srossross — CHANGES_REQUESTED** [review]

File: .github/actions/annotate/index.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                if (!isValidPackageName(packageName)) {
+                    console.log(`::error file=requirements.txt::Invalid package name: ${packageName}`);
+                    continue;
+                }
+                await annotatePackage(packageName);
+            }
+        }
+    } catch (error) {
+        console.log(`::error::Failed to read ${filePath}: ${error.message}`);
+    }
```
Comment (srossross) [inline .github/actions/annotate/index.js]: Please do not hide errors behind log statements: https://en.wikipedia.org/wiki/Error_hiding

it looks like this is the appropriate call:

```
core.setFailed(`Action failed with error ${err}`);
```

File: .github/actions/annotate/index.js (hunk 6 above)
Comment (srossross) [reply .github/actions/annotate/index.js]: opened #97

File: .github/actions/annotate/dist/index.js.LICENSE.txt [hunk 8]
```diff
@@ -0,0 +1,3 @@
+/*! formdata-polyfill. MIT License. Jimmy Wärting <https://jimmy.warting.se/opensource> */
+
+/*! ws. MIT License. Einar Otto Stangvik <einaros@gmail.com> */
```
Comment (srossross) [inline .github/actions/annotate/dist/index.js.LICENSE.txt]: what is this?

File: .github/actions/annotate/index.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+                recommendation = 'The package is in development but poses low risks.';
+            } else if (maturityValue === 'Experimental' || healthRiskValue === 'High Risk') {
+                recommendation = 'This package may pose significant risks to stability and maintainability.';
+            } else if (maturityValue === 'Legacy') {
+                recommendation = 'This package is legacy and may not be stable, consider alternatives.';
+            } else {
+                recommendation = 'Insufficient data to make an informed recommendation.';
+            }
+
+            core.notice(`Package ${packageName}: (Maturity: ${maturityValue}, Health: ${healthRiskValue}). ${recommendation}`);
```
Comment (srossross) [inline .github/actions/annotate/index.js]: There is no output. You need to include the file name and line number

see AnnotationProperties in https://www.npmjs.com/package/@actions/core

File: .github/actions/annotate/dist/index.js.LICENSE.txt (hunk 8 above)
Comment (karamba228) [reply .github/actions/annotate/dist/index.js.LICENSE.txt]: That got auto-filled with running the webpack command. do we have a specific license that we are using? Or should this just be removed completely?

Comment (srossross) [reply .github/actions/annotate/dist/index.js.LICENSE.txt]: this should be removed, but I will add a MIT license in the main dir

Comment (srossross) [reply .github/actions/annotate/dist/index.js.LICENSE.txt]: actually I already set up a BSD 2-clause 

https://github.com/openteamsinc/Score/blob/main/LICENSE

File: .github/actions/annotate/action.yml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+name: "Annotate Requirements"
+description: "Annotates the requirements.txt file using the Score API"
+runs:
+  using: "node20"
+  main: "dist/index.js"
```
Comment (srossross) [inline .github/actions/annotate/action.yml]: it looks like you are missing the intputs here

```suggestion
name: "Annotate Requirements"
description: "Annotates the requirements.txt file using the Score API"
runs:
  using: "node20"
  main: "dist/index.js"

inputs:
  package-ecosystem:
    description: ....
    required: true
```

**Review by srossross — APPROVED** [review]
This looks good, lets move it to https://github.com/openteamsinc/actions/

**Comment by srossross** [comment]
moved to openteamsinc/actions#1

**Pull request closed without merging**